daml-script/dump/src/main/scala/com/daml/script/dump/Dependencies.scala

```diff
@@ -0,0 +1,125 @@
+
+  // Given a list of root package ids, download all packages transitively referenced by those roots.
+  def fetchPackages(client: LedgerClient, references: List[PackageId])(implicit
+      ec: ExecutionContext
+  ): Future[Map[PackageId, (ByteString, Ast.Package)]] = {
+    def go(
+        todo: List[PackageId],
+        acc: Map[PackageId, (ByteString, Ast.Package)],
+    ): Future[Map[PackageId, (ByteString, Ast.Package)]] =
+      todo match {
+        case Nil => Future.successful(acc)
+        case p :: todo if acc.contains(p) => go(todo, acc)
+        case p :: todo =>
+          client.packageClient.getPackage(p).flatMap { pkgResp =>
+            val cos = damlLfCodedInputStreamFromBytes(
+              pkgResp.archivePayload.toByteArray,
+              Decode.PROTOBUF_RECURSION_LIMIT,
+            )
+            val pkgId = PackageId.assertFromString(pkgResp.hash)
+            val pkg = Decode
+              .readArchivePayloadAndVersion(pkgId, DamlLf.ArchivePayload.parser().parseFrom(cos))
+              ._1
+              ._2
+            go(todo ++ pkg.directDeps, acc + (pkgId -> ((pkgResp.archivePayload, pkg))))
+          }
+      }
+    go(references, Map.empty)
+  }
+
+  def writeDar(
+      sdkVersion: String,
+      file: Path,
+      dar: Dar[(PackageId, ByteString, Ast.Package)],
+  ): Unit = {
+    def encode(pkgId: PackageId, bs: ByteString) = {
+      DamlLf.Archive
+        .newBuilder()
+        .setHash(pkgId)
+        .setHashFunction(DamlLf.HashFunction.SHA256)
+        .setPayload(bs)
+        .build()
+    }
+
+    val out = new ZipOutputStream(new FileOutputStream(file.toFile))
+    out.putNextEntry(new ZipEntry("META-INF/MANIFEST.MF"))
+    out.write(manifest(sdkVersion, dar))
+    out.closeEntry()
+    dar.all.foreach { case (pkgId, bs, _) =>
+      out.putNextEntry(new ZipEntry(pkgId + ".dalf"))
+      out.write(encode(pkgId, bs).toByteArray)
+      out.closeEntry
+    }
+    out.close
+  }
+
+  // Given the pkg id of a main dalf and the map of all downloaded packages produce
+  // a DAR or return None for builtin packages like daml-stdlib
+  // that don’t need to be listed in data-dependencies.
+  def toDar(
+      pkgId: PackageId,
+      pkgs: Map[PackageId, (ByteString, Ast.Package)],
+  ): Option[Dar[(PackageId, ByteString, Ast.Package)]] = {
+    def deps(pkgId: PackageId): Set[PackageId] = {
+      @tailrec
+      def go(todo: List[PackageId], acc: Set[PackageId]): Set[PackageId] =
+        todo match {
+          case Nil => acc
+          case p :: todo if acc.contains(p) => go(todo, acc)
+          case p :: todo =>
+            go(todo ++ pkgs(p)._2.directDeps.toList, acc.union(pkgs(p)._2.directDeps))
+        }
+      go(List(pkgId), Set.empty) - pkgId
+    }
+    for {
+      pkg <- pkgs.get(pkgId) if pkg._2.metadata.isDefined
+      if !Seq("daml-stdlib", "daml-prim", "daml-script")
+        .map(Ref.PackageName.assertFromString(_))
+        .contains(pkg._2.metadata.get.name)
+    } yield {
```

Comment:
```suggestion
  }
  
  private val providedLibraries: Set[Ref.PackageName] =
    Set("daml-stdlib", "daml-prim", "daml-script").map(Ref.PackageName.assertFromString(_))

  // Given the pkg id of a main dalf and the map of all downloaded packages produce
  // a DAR or return None for builtin packages like daml-stdlib
  // that don’t need to be listed in data-dependencies.
  def toDar(
      pkgId: PackageId,
      pkgs: Map[PackageId, (ByteString, Ast.Package)],
  ): Option[Dar[(PackageId, ByteString, Ast.Package)]] = {
    def deps(pkgId: PackageId): Set[PackageId] = {
      @tailrec
      def go(todo: List[PackageId], acc: Set[PackageId]): Set[PackageId] =
        todo match {
          case Nil => acc
          case p :: todo if acc.contains(p) => go(todo, acc)
          case p :: todo =>
            go(todo ++ pkgs(p)._2.directDeps.toList, acc.union(pkgs(p)._2.directDeps))
        }
      go(List(pkgId), Set.empty) - pkgId
    }
    for {
      pkg <- pkgs.get(pkgId) if !pkg._2.metadata.exists(providedLibraries)
    } yield {
```

Does it makes sense?

```
scala> val exceptions = Set(1,2,3,4)
exceptions: scala.collection.immutable.Set[Int] = Set(1, 2, 3, 4)

scala> !Some(5).exists(exceptions)
res2: Boolean = true

scala> !Some(1).exists(exceptions)
res3: Boolean = false

scala> !None.exists(exceptions)
res4: Boolean = true
```

Reply:
Sorry, I probably messed up something in the formatting.

Reply:
no worries fixed up in 0b06b97faa32687d9061f524a07d3ea42a6d6ffb, thanks for pointing me to `exists`, that looks much nicer.